parasolr/solr/client.py

```diff
@@ -86,6 +96,81 @@ def _process_facet_counts(self, facet_counts: AttrDict) -> OrderedDict:
         return facet_counts
 
 
+class QueryResponse(BaseResponse):
+    """Thin wrapper to give access to Solr select responses.
+
+    Args:
+        response: A Solr query response
+    """
+
+    def __init__(self, response: Dict) -> None:
+        super().__init__(response)
+        # document list is contained with the "response" element
+        # in the json returned by solr
+        self.document_list = self.response.response
+
+    @property
+    def numFound(self) -> int:
+        return self.document_list.numFound
+
+    @property
+    def start(self) -> int:
+        return self.document_list.start
+
+    @property
+    def docs(self) -> List:
+        return self.document_list.docs
+
+    @property
+    def items(self) -> List:
+        return self.docs
+
+
+class GroupedResponse(BaseResponse):
+    """Query response variant for grouped results.
+
+    Args:
+        response: A Solr query response
+    """
+
+    def __init__(self, response: Dict) -> None:
+        super().__init__(response)
+        # grouped response structure is structured as a dict
+        # first keyed on fieldname with number of matches, then a dict
+        # of group values and corresponding document list
+        self.grouped = self.response.grouped
+
+        # access grouped results at:
+        # self.grouped.fieldname.groups
+        # groups has
+        # will be a dict of field value, doclist
+
+    @property
+    def group_field(self) -> str:
+        "group.field as stored in the params. Not yet supporting grouping by query."
+        return self.params.get("group.field", "")
+
+    @property
+    def numFound(self) -> int:
+        # each field used for grouping has a total
+        # for the number of matches in that grouping
+        # return sum(group["matches"] for group in self.grouped.values())
+        return self.grouped.get(self.group_field, {}).get("matches", 0)
+
+    @property
+    def groups(self) -> List:
+        """Unlike `QueryResponse.docs`, this returns a list of groups with nested documents.
+
+        :return: _description_
+        :rtype: List
+        """
+        return self.grouped.get(self.group_field, {}).get("groups", [])
+
+    @property
+    def items(self) -> List:
+        return self.groups
```

Comment:
Nice abstraction here. I saw the effect of it earlier `.items[k]` instead of `.docs[k]` but it's nice to see it defined here and on QueryResponse.